File: cpp/questions/threeSum.cpp
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        std::sort(nums.begin(), nums.end());
        std::vector<std::vector<int>> ans;
        for (int i=0; i<n-2; ++i) {
            if (nums[i]+nums[i+1]+nums[i+2] > 0) {
                break;
            }
            if ((i>0 && nums[i]==nums[i-1]) || nums[i]+nums[n-2]+nums[n-1]<0) {
                continue;
            }
            int t = nums[i];
            int l = i + 1;
            int r = n - 1;
            while (l < r) {
                int s = t + nums[l] + nums[r];
                if (s > 0) {
                    --r;
                } else if (s < 0) {
                    ++l;
                } else {
                    ans.push_back({t, nums[l], nums[r]});
                    ++l;
                    while (l<r && nums[l-1]==nums[l]) {
                        ++l;
                    }
                    --r;
                    while (l<r && nums[r]==nums[r+1]) {
                        --r;
                    }
                }
            }
        }
        return ans;
    }
};

File: rust/questions/three_sum.rs
impl Solution {
    pub fn three_sum(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let nl = nums.len();
        let mut sn = nums.clone();
        sn.sort();
        let mut ans: Vec<Vec<i32>> = Vec::new();
        let e2 = sn[nl - 2] + sn[nl - 1];
        for i in 0..nl - 2 {
            if sn[i] + sn[i + 1] + sn[i + 2] > 0 {
                break;
            }
            if (i > 0 && sn[i] == sn[i - 1]) || (sn[i] + e2 < 0) {
                continue;
            }
            let mut l = i + 1;
            let mut r = nl - 1;
            while l < r {
                let s = sn[i] + sn[l] + sn[r];
                if s > 0 {
                    r -= 1;
                } else if s < 0 {
                    l += 1;
                } else {
                    ans.push(vec![sn[i], sn[l], sn[r]]);
                    l += 1;
                    while l < r && sn[l] == sn[l - 1] {
                        l += 1;
                    }
                    r -= 1;
                    while l < r && sn[r] == sn[r + 1] {
                        r -= 1;
                    }
                }
            }
        }
        ans
    }
}
